pkg/store/prometheus.go

```diff
@@ -366,27 +396,33 @@ func (p *PrometheusStore) handleStreamedPrometheusResponse(s storepb.Store_Serie
 	var data = p.getBuffer()
 	defer p.putBuffer(data)
 
-	bodySizer := NewBytesRead(httpResp.Body)
-	seriesStats := &storepb.SeriesStatsCounter{}
-
 	// TODO(bwplotka): Put read limit as a flag.
 	stream := remote.NewChunkedReader(bodySizer, remote.DefaultChunkedReadLimit, *data)
 	for {
+		startPromRead := time.Now()
 		res := &prompb.ChunkedReadResponse{}
 		err := stream.NextProto(res)
 		if err == io.EOF {
 			break
 		}
+
 		if err != nil {
 			return errors.Wrap(err, "next proto")
 		}
+		timePrometheusRead += time.Now().Sub(startPromRead).Seconds()
 
 		if len(res.ChunkedSeries) != 1 {
 			level.Warn(p.logger).Log("msg", "Prometheus ReadRequest_STREAMED_XOR_CHUNKS returned non 1 series in frame", "series", len(res.ChunkedSeries))
 		}
 
 		framesNum++
 		for _, series := range res.ChunkedSeries {
+			startSharding := time.Now()
+			if !shardMatcher.MatchesZLabels(series.Labels) {
```

Comment:
Just curious that is the current implementation has better performance than non-sharding queries?
I can see the sidecar will send more remote read requests to Prometheus. Prometheus will respond the same series results back to the sidecar and the filtering is done on the sidecar side. 

This causes more load on the Prometheus side. I am wondering if we can filter on the Prometheus side. Maybe combined with https://github.com/prometheus/prometheus/issues/10420.

Reply:
This is a very good observation, it is hard to shard directly in Prometheus. I did some benchmarking and it seems that the improvements are still there even for Prometheus. I guess people will need to test themselves and we can call this out in the docs, but I believe there is value in sharding in the sidecar.